### PR TITLE
[Kernel][RFC] Initial commit containing new Triton kernels for multi lora serving.

### LINKED ISSUE #2829: [Performance] 40% performance drop using lora vs no lora
```
...
Processed prompts: 100% 1321/1321 [09:09<00:00,  2.40it/s]
training  | ic| "Testing '" + lora + "'": "Testing 'models/v3/checkpoint-1000'"
Processed prompts: 100% 1321/1321 [16:01<00:00,  1.37it/s]
training  | ic| "Lora '" + lora + "' testing complete.": "Lora 'models/v3/checkpoint-1000' testing complete."
training  | ic| "Testing '" + lora + "'": "Testing 'models/v3/final_lora'"
Processed prompts: 100% 1321/1321 [15:59<00:00,  1.38it/s]
training  | ic| "Lora '" + lora + "' testing complete.": "Lora 'models/v3/final_lora' testing complete."
training  | ic| "Testing '" + lora + "'": "Testing 'models/v3/checkpoint-2000'"
...
```

Testing offline batched inference with a big bunch of long context, short reply prompts, feeding the vllm engine all 1321 prompts as a list in a single batch. Tested on an 80GB A100.

This isn't a bug per se, or a major problem, more like a... note. I guess a big performance drop for loras is expected?

### PR DESCRIPTION
**Pull request by FurtherAI**
# SGMV Triton Kernels
Should Fix #2829, #4007, #4053, #4063, #3793, #4708

Modification of #5025 to change to contiguous weight format rather than S-LoRA paged weights.
In parallel with #5036

---

New Triton kernels for multi lora computation. These (should) handle any shape and data type, compute at the actual lora rank and also speed up for grouped lora requests (especially prefill).

- [ ] Replace Punica kernels.

ping @Yard1

### CODE REVIEW

**Comment by robertgshaw2-neuralmagic** [comment]
@tlrmchlsmth

File: benchmarks/kernels/benchmark_sgmv_triton.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
+import torch
+import triton
+
+from tests.lora.test_sgmv_triton import MAX_TEST_POWER, setup_
```
Comment (tlrmchlsmth) [inline benchmarks/kernels/benchmark_sgmv_triton.py:4]: Ideally, benchmark code shouldn't depend on the `tests` module -- I had to manually set my PYTHONPATH to run this. Duplicating the `setup_` function might be a better way to do it in this case, as much as I hate to suggest it.

File: benchmarks/kernels/benchmark_sgmv_triton.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+from vllm.model_executor.layers.lora import sgmv_triton as sgmv
+
+
+@triton.testing.perf_report(
```
Comment (tlrmchlsmth) [inline benchmarks/kernels/benchmark_sgmv_triton.py:8]: Looks like triton testing needs both `pandas` and `matplotlib` -- could you add those to `requirements-dev.txt` in the benchmarking section?

**Review by tlrmchlsmth — COMMENTED** [review]
Just starting to take a look -- Could you explain a bit about the change to a contiguous weight format?

File: benchmarks/kernels/benchmark_sgmv_triton.py (hunk 1 above)
Comment (FurtherAI) [reply benchmarks/kernels/benchmark_sgmv_triton.py:4]: I could put the function somewhere in main vllm rather than in the test module

File: benchmarks/kernels/benchmark_sgmv_triton.py (hunk 2 above)
Comment (FurtherAI) [reply benchmarks/kernels/benchmark_sgmv_triton.py:8]: Sure

**Comment by FurtherAI** [comment]
@tlrmchlsmth So, S-LoRA puts the weights in a page that is `[page_size, hidden_dim]`, each LoRA matrix has `rank` weights there. By contiguous, I mean going back to the format that is currently implemented for the Punica kernels, where the LoRA weights are stored in a stack that is padded to max rank (e.g. `[max_loras, 1, hidden_dim, max_rank]` for LoRA B)`.

**Review by Yard1 — APPROVED** [review]
This looks good - I think we can just merge it. Merging master should fix CI.

**Comment by FurtherAI** [comment]
@Yard1 You don't mean merge as in finish this PR already right? We still gotta remove punica

**Comment by Yard1** [comment]
@FurtherAI We can merge the kernels first, and then follow up with a PR to enable them. But either works!

**Comment by jeejeelee** [comment]
@Yard1  I'm still waiting for your decision, but you haven't provided any feedback, so I don't quite understand.

**Comment by jeejeelee** [comment]
@robertgshaw2-neuralmagic  @tlrmchlsmth  I would appreciate it if you could pay attention to my related PR.

**Comment by FurtherAI** [comment]
@Yard1 @tlrmchlsmth @robertgshaw2-neuralmagic 
So, I am working on removing Punica and I think I've written all the code for it, working on testing now. Appreciate any pointers if anyone has any ideas about the errors I'm running into.

The first time I run a layer (currently testing column parallel in `test_layers.py`), it works. After that, it has a particular order of working and not working (seed is set to the same value at all times), being either off by 0.2 or 500 on average. This happens whether the layer is run in a loop or if the function is duplicated multiple times (they both repeat the same sequence). Pytest, Torch and Triton (autotuner disabled) all might persist something across different function calls, but I am baffled by this behavior, since I think they should be independent. Has anyone seen this before or have any ideas?

I pushed a commit with the current state.

**Comment by FurtherAI** [comment]
Looks like I need Triton >= 2.2.0. Where all should the requirements be updated for this?

**Comment by chandan047** [comment]
Hi, our team is really looking forward to this merge! Is there an estimate on when this will be merged?

**Comment by FurtherAI** [comment]
@chandan047 This was largely taken over by #5036, though these kernels have the ability to compute at the actual rank of each LoRA, which I'm not sure was added in the other PR. They are also simpler, but I am not atm planning to try to merge this with the updated main. The other ones were merged into main so see if that works for your use case
